File: week-11/pushedrumex/사라지는 발판.py
answer = int(1e9)
def solution(board, aloc, bloc):
    play(board, aloc, bloc, 0)
    return answer

dxdy = ((0,1),(1,0),(-1,0),(0,-1))
def play(board, aloc, bloc, count):
    global answer
    ax, ay = aloc
    bx, by = bloc
    if board[ax][ay] == 0 or board[bx][by] == 0:
        answer = min(answer, count)

    flag = True
    if count % 2 == 0:
        for dx, dy in dxdy:
            _x = ax + dx
            _y = ay + dy
            if not (0 <= _x < len(board) and 0 <= _y < len(board[0])): continue
            if board[_x][_y] == 0: continue
            flag = False
            board[ax][ay] = 0
            play(board, (_x, _y), bloc, count + 1)
            board[ax][ay] = 1
    else:
        for dx, dy in dxdy:
            _x = bx + dx
            _y = by + dy
            if not (0 <= _x < len(board) and 0 <= _y < len(board[0])): continue
            if board[_x][_y] == 0: continue
            flag = False
            board[bx][by] = 0
            play(board, aloc, (_x, _y), count + 1)
            board[bx][by] = 1
    if flag:
        answer = min(answer, count)

File: week-11/dahyen0o/사라지는-발판.java
import java.util.*;

class Solution {
    
    private static final int[] dr = {0, -1, 0, +1};
    private static final int[] dc = {-1, 0, +1, 0};
    
    private static int answer = 0;
    
    public int solution(final int[][] board, int[] A, int[] B) {
        final int[] result = play(0, board, A, B);
        
        return result[1];
    }
    
    private int[] play(final int round, final int[][] board, int[] A, int[] B) {
        // player 기준 최선의 선택을 반환
        // {게임 결과, 최선의 round}
        final int[] player = (round % 2 == 0) ? A : B;
        
        if(board[player[0]][player[1]] == 0) {
            return new int[] {0, round};
        }
        
        final List<int[]> results = new ArrayList<>();
        
        for(int dir = 0; dir < 4; dir++) {
            final int r = player[0] + dr[dir];
            final int c = player[1] + dc[dir];
            if(r < 0 || c < 0 || r >= board.length || c >= board[0].length) continue;
            if(board[r][c] == 0) continue;
            
            board[player[0]][player[1]] = 0;
            final int[] result = (round % 2 == 0) 
                ? play(round + 1, board, new int[] {r, c}, B) 
                : play(round + 1, board, A, new int[] {r, c});
            
            results.add(result);
            board[player[0]][player[1]] = 1;
        }
        
        if(results.size() == 0) return new int[] {0, round};
        
        int win = Integer.MAX_VALUE, lose = 0;
        for(final int[] result : results) {
            if(result[0] == 0) { // 상대방 lose
                win = Math.min(win, result[1]);
            }
            else { // 상대방 win
                lose = Math.max(lose, result[1]);
            }
        }
        return (win == Integer.MAX_VALUE) ? new int[] {0, lose} : new int[] {1, win};
    }
}
